src/main/scala/mesosphere/marathon/core/health/impl/HealthCheckActor.scala

```diff
@@ -15,113 +16,76 @@ import mesosphere.marathon.core.task.termination.{ KillReason, KillService }
 
+import scala.collection.concurrent.TrieMap
 import scala.concurrent.Future
 import scala.concurrent.duration._
 import scala.util.{ Failure, Success }
 
 private[health] class HealthCheckActor(
-    app: AppDefinition,
-    appHealthCheckActor: ActorRef,
-    killService: KillService,
-    healthCheck: HealthCheck,
-    instanceTracker: InstanceTracker,
-    eventBus: EventStream)(implicit mat: Materializer) extends Actor with StrictLogging {
+  app: AppDefinition,
+  appHealthCheckActor: ActorRef,
+  killService: KillService,
+  healthCheck: HealthCheck,
+  instanceTracker: InstanceTracker,
+  eventBus: EventStream,
+  healthCheckHub: Sink[(AppDefinition, Instance, MarathonHealthCheck, ActorRef), NotUsed])(implicit mat: ActorMaterializer)
+    extends Actor with StrictLogging {
 
-  import HealthCheckWorker.HealthCheckJob
   import context.dispatcher
 
-  var nextScheduledCheck: Option[Cancellable] = None
-  var healthByInstanceId = Map.empty[Instance.Id, Health]
-
-  val workerProps: Props = Props(classOf[HealthCheckWorkerActor], mat)
+  var healthByInstanceId = TrieMap.empty[Instance.Id, Health]
 
   override def preStart(): Unit = {
-    logger.info(
-      "Starting health check actor for app [{}] version [{}] and healthCheck [{}]",
-      app.id,
-      app.version,
-      healthCheck
-    )
-    //Start health checking not after the default first health check
-    val start = math.min(healthCheck.interval.toMillis, HealthCheck.DefaultFirstHealthCheckAfter.toMillis).millis
-    scheduleNextHealthCheck(Some(start))
-  }
-
-  override def preRestart(reason: Throwable, message: Option[Any]): Unit =
-    logger.info(
-      "Restarting health check actor for app [{}] version [{}] and healthCheck [{}]",
-      app.id,
-      app.version,
-      healthCheck
-    )
-
-  override def postStop(): Unit = {
-    nextScheduledCheck.forall { _.cancel() }
-    logger.info(
-      "Stopped health check actor for app [{}] version [{}] and healthCheck [{}]",
-      app.id,
-      app.version,
-      healthCheck
-    )
-  }
-
-  def updateInstances(): Future[Done] = {
-    instanceTracker.specInstances(app.id).map { instances =>
-      self ! InstancesUpdate(version = app.version, instances = instances)
-      Done
-    }.recover {
-      case t =>
-        logger.error(s"An error has occurred: ${t.getMessage}", t)
-        Done
+    healthCheck match {
+      case marathonHealthCheck: MarathonHealthCheck =>
+        //Start health checking not after the default first health check
+        val startAfter = math.min(marathonHealthCheck.interval.toMillis, HealthCheck.DefaultFirstHealthCheckAfter.toMillis).millis
+
+        logger.info(s"Starting health check for ${app.id} version ${app.version} and healthCheck $marathonHealthCheck in $startAfter ms")
+
+        Source
+          .tick(startAfter, marathonHealthCheck.interval, Tick)
+          .mapAsync(1)(_ => instanceTracker.specInstances(app.id))
```

Comment:
Come to think of it we really should not fetch all instances each tick. Anyhow, that's for another time.